### DISCUSSION TITLE
Loading HLS/Liquidsoap through https port 9443

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3844] [User Support]
Hello everyone.
I am slowly pushing along to get this up and running.
I created and installed Let's Encrypt SSL into Nginx last night.
Port 9443 enabled and pointing to the Linux server 192.168.5.10 on the **PFSense** Nat/Rule.

I have the following in my Config files.
I've placed this in the following areas.
```
/usr/local/nginx/conf/nginx.conf
/etc/nginx/sites-available/stream.example.com
```
```
    server {
      listen 443 ssl;
      listen [::]:443 ssl;

      server_name stream.example.com;
      ssl_certificate /etc/letsencrypt/live/stream.example.com//fullchain.pem;
      ssl_certificate_key /etc/letsencrypt/live/stream.example.com//privkey.pem;

      # Redirect the browser to our port 9443 config
      return 301 $scheme://stream.example.com/:9443$request_uri;
    }

    server {
      listen 9443 ssl;
      listen [::]:9443 ssl;

      server_name stream.example.com/;
      ssl_certificate /etc/letsencrypt/live/stream.example.com//fullchain.pem;
      ssl_certificate_key /etc/letsencrypt/live/stream.example.com//privkey.pem;
      add_header Strict-Transport-Security "max-age=31536000; includeSubDomains";

      #Configures the publicly served root directory
      #Configures the index file to be served
      root /var/www/stream.example.com/;
      index index.html index.htm;
}
```

The webpage to the Linux server loads from url 'https://stream.example.com:9443'
The webpage loads with SSL:9443 and gives a valid certificate.

The issue now is loading the m3u8 file under https port 9443 from either **inside** or **outside** the network. 
When I try it in Chrome, I get a 404 Not Found error instead of the download dialog.

When I try it in Windows Media Player, I get "Failed to render the file."
The URL is (Example URL) `https://stream.example.com:9443/hls/radio101.m3u8`

The error in Chrome
```
404 Not Found
nginx/1.18.0 (Ubuntu)
```
Any ideas on that one?

**Comment by vitoyucepi** [comment]
Hi @carrzkiss,
I don't know if you have the `stream.example.com` enabled correctly, but I assume it's fine.
This line should also look like this
```diff
-return 301 $scheme://stream.example.com/:9443$request_uri;
+return 301 https://stream.example.com:9443$request_uri;
```

According to the config
```nginx
root /var/www/stream.example.com/;
index index.html index.htm;
```
the url `https://stream.example.com:9443/hls/radio101.m3u8` serves the
`/var/www/stream.example.com/hls/radio101.m3u8` file. If the file is located elsewhere, you should create a location section.
```nginx
root /var/www/stream.example.com/;
index index.html index.htm;
location /hls/ {
  root /path/to/hls/;
}
```
Alternatively, you could move the files generated by liquidsoap to `output.file.hls("/var/www/stream.example.com/hls", streams, radio101)`.
You can also simlink directories.

**Reply by carrzkiss** [reply]
Hey, Vito.
Changed to this.
`return 301 https://stream.example.com:9443$request_uri;`
(Changed in both location, nginx.conf and sites-available/)

Also, I  added the **location**.
```
  root /var/www/stream.example.com/;
  index index.html index.htm;
  location /hls/ {
  root /var/www/html/hls/;
```
The **location** is from 101.liq

`hls_path /var/www/html/hls/;`

After making all the changes I performed
```
sudo fuser -k 80/tcp
sudo fuser -k 9443/tcp
sudo systemctl restart nginx
sudo systemctl restart liquidsoap@101
sudo /usr/local/nginx/sbin/nginx -g 'daemon off;'
```

But now, the local does not work.
So, I have to find out what I broke.
running this `sudo systemctl status liquidsoap@101.service`
Shows that it is `Active: active (running) `
But, it will not run in a media player or cause a file download in Chrome for the local test, `http://192.168.5.10/hls/radio101.m3u8`.

I will post back once I figure out what happened.

**Reply by carrzkiss** [reply]
Reboot fixed the issue with local playing.
I still get the `404 not found` using the domain URL.

The SSL is working on the domain:9443 address, so that is good. It is just not loading the download dialog in Chrome and still giving the same error in Media Player.

**Reply by vitoyucepi** [reply]
Try this instead.
```nginx
location /hls/ {
  root /var/www/html;
}
```
I forgot to mention that `/hls/` in location will be the part of the path.
https://nginx.org/en/docs/beginners_guide.html#static
When you request `http://192.168.5.10/hls/radio101.m3u8` the server will send `/var/www/html/hls/radio101.m3u8`.


These commands look nasty.
```
sudo fuser -k 80/tcp
sudo fuser -k 9443/tcp
sudo systemctl restart nginx
sudo systemctl restart liquidsoap@101
sudo /usr/local/nginx/sbin/nginx -g 'daemon off;'.
```
However, after `make install`, the traditional methods may not work reliably.

**Comment by carrzkiss** [comment]
OK, have it like this in both locations.
```
  root /var/www/stream.cffradio.com/;
  index index.html index.htm;
  location /hls/ {
     root /var/www/html;
}
```
It will only work locally, not domain.

When I run this.
`sudo /usr/local/nginx/sbin/nginx -g 'daemon off;'`
I have to stop it on port 80 and port 9443.
Which is these are there.
```
sudo fuser -k 80/tcp
sudo fuser -k 9443/tcp
```

Do I need to run `sudo /usr/local/nginx/sbin/nginx -g 'daemon off;'` now or not?

**Reply by vitoyucepi** [reply]
It's very difficult for me to predict exactly what the problem might be.
Ideally, you should just
1. install nginx from the repository;
1. enable and start the default config with `systemctl enable --now nginx`;
1. check that the http server is working on the ip;
1. create a new file in `/etc/nginx/sites-available/site.conf`;
1. symlink file from `/etc/nginx/sites-available/site.conf` to `/etc/nginx/sites-enabled/site.conf`;
1. test the configuration with `nginx -t`;
1. reload server configuration with `systemctl reload nginx`;
1. check if everything works;
1. if something is wrong, edit `/ect/nginx/sites-available/site.conf`, `nginx -t` and `systemctl reload nginx`;
1. in some cases, such as changing ports, `systemctl restart nginx` is required.

**Comment by carrzkiss** [comment]
IT JUST STARTED WORKING!!!

I was working on the website and closing out of tabs when I came across the domain URL tab and decided to refresh and It loaded the download dialog.
I placed the URL in Windows Media Player Classic and am now listening to the live domain URL radio.

I don't know what the deal was and why it took so long, but it is working now.
This is crazy, but it's working.

Sadly, no metadata is being passed.
I will have to look at the link I shared with Yaw the other day and see if that will grab it with HLS.
I think he only developed it for Icecast. I will have to contact him to see if he has tried it with HLS.

Thank you again, Vito.
You have an amazing week.

**Reply by toots** [reply]
Glad to hear! 

Metadata in HLS is an on-going work. It should work fine on our end when receiving `mpegts` encapsulated data. If you're using plain `adts` (for instance mp3 or aac), FFmpeg is not currently able to decode in-stream metadata. There's a patch pending: https://ffmpeg.org/pipermail/ffmpeg-devel/2024-March/324312.html

**Reply by toots** [reply]
As for other players, I have no idea.. 🤷
